# PRUEBA 2 (TODO LO APRENDIDO ENTRE CAPS 5-10)
Temario:

1- control de flujo

2- filtrar data, manejo de duplicados

3- extraccion y actualizacion de datos puntuales

4- tablas pivote con funciones agregadas

5- Gráfico estatico e interactiva

6- funcion group by, agg y lambda

7- concatenacion dataframes (left, inner y full joins)

# 1. CONTROL DE FLUJO

1.1 Genere un dataframe de 15 nombres  y sus generos, con notas aleatorias siguiendo una normal con media en 5 y desviacion estandar 1, esto para notas de castellano, ciencias y matematicas

1.2 Entregue un dataframe solo con aquellos que tienen sobre 6.1 en matematicas (usando for e if)

1.3 Genere una nueva columna donde salga el promedio general de todos, matematicas pondera 50%, ciencias 20% y castellano 30%

1.4 recorra con for el dataframe y ponga "aprobado" o "reprobrado" si el promedio es menor a 4.5

1.5 ordene de mejor nota a peor nota promedio final

In [1]:
#1.1
import pandas as pd 
import numpy as np 
nombres=['Diego','Ruben','Maria Jose','Daniel','Damian','Constanza','Hernan','Javiera','Francisca','Sergio','Macarena','Roberto','Andrea','Maria Paz','Jose']
apellidos=['Bertin','Bertin','Silva','Bertin','Bertin','Ljubetic','Olmedo','Gonzalez','Molina','Osorio','Osorio','Zahr','Gonzalez','Lagos','Vergara']
genero=['M','M','F','M','M','F','M','F','F','M','F','M','F','F','M']
DICC_NOTAS={'NOMBRES':nombres,'APELLIDOS':apellidos,'GENERO':genero}
DF_NOTAS=pd.DataFrame(DICC_NOTAS)
DF_NOTAS['CASTELLANO']=np.round(np.clip(np.random.normal(5.5,2,len(DF_NOTAS)),1,7), decimals=1)
DF_NOTAS['MATEMATICAS']=np.round(np.clip(np.random.normal(6,1.1,len(DF_NOTAS)),1,7), decimals=1)
DF_NOTAS['CIENCIAS']=np.round(np.clip(np.random.normal(5,2,len(DF_NOTAS)),1,7), decimals=1)
DF_NOTAS

NOMBRES APELLIDOS GENERO  CASTELLANO  MATEMATICAS  CIENCIAS
0        Diego    Bertin      M         7.0          6.0       2.6
1        Ruben    Bertin      M         7.0          6.3       3.2
2   Maria Jose     Silva      F         5.5          7.0       2.4
3       Daniel    Bertin      M         4.2          6.5       7.0
4       Damian    Bertin      M         6.0          5.5       3.3
5    Constanza  Ljubetic      F         7.0          4.9       3.6
6       Hernan    Olmedo      M         4.4          5.8       4.3
7      Javiera  Gonzalez      F         6.3          5.5       4.0
8    Francisca    Molina      F         5.7          7.0       7.0
9       Sergio    Osorio      M         5.7          5.8       4.4
10    Macarena    Osorio      F         3.7          6.0       3.8
11     Roberto      Zahr      M         7.0          5.4       2.4
12      Andrea  Gonzalez      F         5.8          7.0       4.7
13   Maria Paz     Lagos      F         7.0          5.4       5.0
14        Jose   Vergara      M         3.9          4.8       4.4

In [2]:
DF_NOTAS.iloc[3:4,:]

NOMBRES APELLIDOS GENERO  CASTELLANO  MATEMATICAS  CIENCIAS
3  Daniel    Bertin      M         3.7          6.0       4.6

In [3]:
DF_NOTAS.iloc[3,:]

NOMBRES        Daniel
APELLIDOS      Bertin
GENERO              M
CASTELLANO        3.7
MATEMATICAS       6.0
CIENCIAS          4.6
Name: 3, dtype: object

In [2]:
#1.2
df_MATEMATICAS = pd.DataFrame()
for i in range(len(DF_NOTAS)):
    if DF_NOTAS.iloc[i,4]>=6.1:   
        df_MATEMATICAS = pd.concat([df_MATEMATICAS, DF_NOTAS.iloc[i:i+1,:]])
df_MATEMATICAS.reset_index(drop=True, inplace=True)
df_MATEMATICAS

NOMBRES APELLIDOS GENERO  CASTELLANO  MATEMATICAS  CIENCIAS
0       Ruben    Bertin      M         7.0          6.3       3.2
1  Maria Jose     Silva      F         5.5          7.0       2.4
2      Daniel    Bertin      M         4.2          6.5       7.0
3   Francisca    Molina      F         5.7          7.0       7.0
4      Andrea  Gonzalez      F         5.8          7.0       4.7

In [3]:
# 1.3 nueva columna

for i in range(len(DF_NOTAS)):  
    DF_NOTAS.loc[i,'PROMEDIO']=0.5*DF_NOTAS.loc[i,'MATEMATICAS']+0.2*DF_NOTAS.loc[i,'CIENCIAS']+0.3*DF_NOTAS.loc[i,'CASTELLANO']
DF_NOTAS


NOMBRES APELLIDOS GENERO  CASTELLANO  MATEMATICAS  CIENCIAS  PROMEDIO
0        Diego    Bertin      M         7.0          6.0       2.6      5.62
1        Ruben    Bertin      M         7.0          6.3       3.2      5.89
2   Maria Jose     Silva      F         5.5          7.0       2.4      5.63
3       Daniel    Bertin      M         4.2          6.5       7.0      5.91
4       Damian    Bertin      M         6.0          5.5       3.3      5.21
5    Constanza  Ljubetic      F         7.0          4.9       3.6      5.27
6       Hernan    Olmedo      M         4.4          5.8       4.3      5.08
7      Javiera  Gonzalez      F         6.3          5.5       4.0      5.44
8    Francisca    Molina      F         5.7          7.0       7.0      6.61
9       Sergio    Osorio      M         5.7          5.8       4.4      5.49
10    Macarena    Osorio      F         3.7          6.0       3.8      4.87
11     Roberto      Zahr      M         7.0          5.4       2.4      5.28
12      Andrea  Gonzalez      F         5.8          7.0       4.7      6.18
13   Maria Paz     Lagos      F         7.0          5.4       5.0      5.80
14        Jose   Vergara      M         3.9          4.8       4.4      4.45

In [4]:
#1.4
for i in range(len(DF_NOTAS)):  
    if DF_NOTAS.loc[i,'PROMEDIO']>=4.5:
        DF_NOTAS.loc[i,'STATUS']='APROBADO'
    else:
        DF_NOTAS.loc[i,'STATUS']='REPROBADO'
DF_NOTAS

NOMBRES APELLIDOS GENERO  CASTELLANO  MATEMATICAS  CIENCIAS  PROMEDIO  \
0        Diego    Bertin      M         7.0          6.0       2.6      5.62   
1        Ruben    Bertin      M         7.0          6.3       3.2      5.89   
2   Maria Jose     Silva      F         5.5          7.0       2.4      5.63   
3       Daniel    Bertin      M         4.2          6.5       7.0      5.91   
4       Damian    Bertin      M         6.0          5.5       3.3      5.21   
5    Constanza  Ljubetic      F         7.0          4.9       3.6      5.27   
6       Hernan    Olmedo      M         4.4          5.8       4.3      5.08   
7      Javiera  Gonzalez      F         6.3          5.5       4.0      5.44   
8    Francisca    Molina      F         5.7          7.0       7.0      6.61   
9       Sergio    Osorio      M         5.7          5.8       4.4      5.49   
10    Macarena    Osorio      F         3.7          6.0       3.8      4.87   
11     Roberto      Zahr      M         7.0          5.4       2.4      5.28   
12      Andrea  Gonzalez      F         5.8          7.0       4.7      6.18   
13   Maria Paz     Lagos      F         7.0          5.4       5.0      5.80   
14        Jose   Vergara      M         3.9          4.8       4.4      4.45   

       STATUS  
0    APROBADO  
1    APROBADO  
2    APROBADO  
3    APROBADO  
4    APROBADO  
5    APROBADO  
6    APROBADO  
7    APROBADO  
8    APROBADO  
9    APROBADO  
10   APROBADO  
11   APROBADO  
12   APROBADO  
13   APROBADO  
14  REPROBADO

In [5]:
#1.5
DF_NOTAS.sort_values(by= 'PROMEDIO',ascending=False)
DF_NOTAS.reset_index()

index     NOMBRES APELLIDOS GENERO  CASTELLANO  MATEMATICAS  CIENCIAS  \
0       0       Diego    Bertin      M         7.0          6.0       2.6   
1       1       Ruben    Bertin      M         7.0          6.3       3.2   
2       2  Maria Jose     Silva      F         5.5          7.0       2.4   
3       3      Daniel    Bertin      M         4.2          6.5       7.0   
4       4      Damian    Bertin      M         6.0          5.5       3.3   
5       5   Constanza  Ljubetic      F         7.0          4.9       3.6   
6       6      Hernan    Olmedo      M         4.4          5.8       4.3   
7       7     Javiera  Gonzalez      F         6.3          5.5       4.0   
8       8   Francisca    Molina      F         5.7          7.0       7.0   
9       9      Sergio    Osorio      M         5.7          5.8       4.4   
10     10    Macarena    Osorio      F         3.7          6.0       3.8   
11     11     Roberto      Zahr      M         7.0          5.4       2.4   
12     12      Andrea  Gonzalez      F         5.8          7.0       4.7   
13     13   Maria Paz     Lagos      F         7.0          5.4       5.0   
14     14        Jose   Vergara      M         3.9          4.8       4.4   

    PROMEDIO     STATUS  
0       5.62   APROBADO  
1       5.89   APROBADO  
2       5.63   APROBADO  
3       5.91   APROBADO  
4       5.21   APROBADO  
5       5.27   APROBADO  
6       5.08   APROBADO  
7       5.44   APROBADO  
8       6.61   APROBADO  
9       5.49   APROBADO  
10      4.87   APROBADO  
11      5.28   APROBADO  
12      6.18   APROBADO  
13      5.80   APROBADO  
14      4.45  REPROBADO

In [10]:

# bonus, crear una columna con notas de musica, los primeros 7 del index con media 5 y los otros 8 con media 3
j=6
for i in range(len(DF_NOTAS)):
    if i<=j: 
        DF_NOTAS.loc[i,'MUSICA']=np.round(np.clip(np.random.normal(6,1,1),1,7), decimals=1)
    else: 
        DF_NOTAS.loc[i,'MUSICA']=np.round(np.clip(np.random.normal(3,2,1),1,7), decimals=1)
DF_NOTAS

NOMBRES APELLIDOS GENERO  CASTELLANO  MATEMATICAS  CIENCIAS  PROMEDIO  \
0        Diego    Bertin      M         7.0          6.0       2.6      5.62   
1        Ruben    Bertin      M         7.0          6.3       3.2      5.89   
2   Maria Jose     Silva      F         5.5          7.0       2.4      5.63   
3       Daniel    Bertin      M         4.2          6.5       7.0      5.91   
4       Damian    Bertin      M         6.0          5.5       3.3      5.21   
5    Constanza  Ljubetic      F         7.0          4.9       3.6      5.27   
6       Hernan    Olmedo      M         4.4          5.8       4.3      5.08   
7      Javiera  Gonzalez      F         6.3          5.5       4.0      5.44   
8    Francisca    Molina      F         5.7          7.0       7.0      6.61   
9       Sergio    Osorio      M         5.7          5.8       4.4      5.49   
10    Macarena    Osorio      F         3.7          6.0       3.8      4.87   
11     Roberto      Zahr      M         7.0          5.4       2.4      5.28   
12      Andrea  Gonzalez      F         5.8          7.0       4.7      6.18   
13   Maria Paz     Lagos      F         7.0          5.4       5.0      5.80   
14        Jose   Vergara      M         3.9          4.8       4.4      4.45   

       STATUS  MUSICA  
0    APROBADO     5.4  
1    APROBADO     7.0  
2    APROBADO     5.2  
3    APROBADO     5.8  
4    APROBADO     7.0  
5    APROBADO     5.1  
6    APROBADO     5.6  
7    APROBADO     5.1  
8    APROBADO     1.6  
9    APROBADO     1.0  
10   APROBADO     5.0  
11   APROBADO     4.3  
12   APROBADO     2.5  
13   APROBADO     1.0  
14  REPROBADO     2.8

# 2. FILTRAR DATA

2.1 Del dataframe anterior (creado en 1.1), filtre aquellos que tienen sobre un 6.1 en matematicas y un 5.0 en lenguaje

2.2 Del dataframe anterior (creado en 1.1), filtre aquellos que tienen sobre un 6.1 en matematicas o un 5.0 en lenguaje

2.3 Cree una columna donde promedio final=0.5 * nota matematicas + 0.3 * nota castellano + 0.2 * nota ciencias y cree un criterio donde si promedio final >4.5 aprobado, entre 4.5 y 3.5 repechaje y bajo 3.5 reprobrado

2.4 usando isin, cree una lista con notas entre 5.5 y 6.3 y vea quienes estan ¿cuantos son?

2.5 Elija 4 estudiantes al azar y que compartan apellido, luego encuentrelos usando el metodo duplicates

2.6 muestre solo un hermano, utilizando duplicated()

2.7 repita, haciendo drop_duplicates

2.8 usando el metodo unique y nunique, cuente valores unicos


# 3. Extracción y actualización de datos puntuales

Vuelva el nombre de los alumnos el indice mediante el metodo de fijar index

3.1 detecte el alumno en el indice 10, y pongale 5 en matematicas, 6 en ciencia y 7 en lenguaje (loc)

3.2 elija a un hermano, y con iloc actualice sus notas

3.3 recalcule con estos cambios la columna promedio con el metodo apply y lambda

3.4 con el metodo query filtre el dataframe de aquellos alumnos que sean hombres y tengan >5 en matematicas

3.5 con sample cree un dataframe y vuelva a aplicar sobre este el >5

# 4. tablas pivote con funciones agregadas

4.1 cree un data frame de largo 1000, en la primera columna ponga un numero que siga una normal de promedio 500 y desv 200, que su minimo sea 100

4.2 cree una segunda fila que multiplique ese valor por una normal de promedio 1 con desv estandar 0,05

4.3 Si la columna 1 es mayor a la columna 2, ponga " 1 gana" si es menor "2 gana" si son iguales " empate", pongale a esta columna "criterio"

4.4 haga una tabla pivote por criterio y use aggfunc="sum"

# 5. Gráfico estatico e interactivo

5.1 haga un diagrama de dispersion de la columna del dataframe anterior, que sea estatico

5.2 haga lo mismo pero dinamico

# 6. group by y agg

Ejecute este codigo
import pandas as pd 
df_cars=pd.read_csv(r'C:\Users\Usuario\Desktop\python\Car_sales.csv')
df_cars=df_cars[['Manufacturer','Sales_in_thousands','Vehicle_type','Price_in_thousands','Engine_size','Horsepower','Fuel_capacity']]

6.1 haga un group by por 'Vehicle_type' eligiendo estas columnas 'Sales_in_thousands','Price_in_thousands'

6.2 usando group by sin funcion, hagalo sobre Manufacturer y muestre en el dataframe solo las ford

6.3 muestre el promedio de los valores numericos para estas marcas 'Toyota','Dodge','Jeep','Honda','Ford'
 
6.4 filtre por esas marcas y vea en vehicle type cuantos son car o passenger

6.5 usando group by y agg calcule el minimo price un thousands y el max horse power para los vehicle type, explique que esta mostrando

6.6 intente hacer el punto 6.5 solo con group by, diga que pasa?

6.7 usando funcion lambda y group by calcule el ratio entre engine size y horsepower, ordene de mayor a menor, esto para las marcas y luego para el tipo de vehiculo

6.8 con un grafico de barras estatico, muestre el resultado ordenado de mayor a menor

6.9 usando filter, filtre por manufacturer aquellos que suman mas de 52 en sales_in_thousands

# 7. Concatenacion dataframes (left, inner y full joins)

7.1 con un random entre 1 y 10000 haga un dataframe de 1000 lineas, copie la columna y vuelvala el index

7.2 repita con otros datos, pero el dataframe hagalo de 800 lineas

7.3 haga un full outer join

7.4 haga un inner join

7.5 haga un left join

7.6 haga un left join exclusivo